portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java

```diff
@@ -75,16 +89,65 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         /** Don't set limit if not needed. */
     }
 
-    /** Use query to populate a list of Comment objects. */
+    /* Use query to populate a list of Comment objects. */
     ArrayList<Comment> comments = new ArrayList<Comment>();
-    
-    for (Entity entity : commentResults.asIterable(options)) {
-        long id = entity.getKey().getId();
-        String message = (String) entity.getProperty("message");
-        long timestamp = (long) entity.getProperty("timestamp");
+    ArrayList<Entity> commentEntities = new ArrayList<Entity>();
+
+    /* Add entities to a list for convenient access later */
+    for (Entity entity: commentResults.asIterable(options)) {
+      commentEntities.add(entity);
+    }
+
+    /* Create a threadpool to perform multiple translations in parallel */
+    ExecutorService executor = Executors.newFixedThreadPool(10);
+    /* A list to hold pending translations in the order they were added. */
+    CompletableFuture<String>[] translationFutures = new CompletableFuture[commentEntities.size()];
+
+    /* Iterate entities and start a translation async for each */
+    for (int i = 0; i < commentEntities.size(); i++) {
+      Entity entity = commentEntities.get(i);
+      
+      /* Lambda expression that captures the entity, performs translation, returns translated message. */
+      Supplier<String> translationTask = () -> {
+        String message = (String) entity.getProperty("message"); 
+        
+        /* A new translation service is created each time because Translate may not be threadsafe.*/
+        Translate translate = TranslateOptions.getDefaultInstance().getService();
+        Translation translation = translate.translate(
+          message,
+          Translate.TranslateOption.targetLanguage(languageCode == null ? "en" : languageCode)
+        );
+        return translation.getTranslatedText();
+      };
+
+      translationFutures[i] = CompletableFuture.supplyAsync(translationTask, executor);
+    }
 
-        Comment comment = new Comment(id, message, projectId, timestamp);
-        comments.add(comment);
+    /* Block until all translations are complete.
+     * This isn't ideal but I'm willing to bet traffic isn't a huge problem. 
+     * For this site, fast load for one >>> handle many users.
+     */
+    CompletableFuture.allOf(translationFutures).join();
```

Comment:
Happy to chat about threading, response times and server throughput if you're interested.  I'm not the most knowledgeable on this specific platform but I assume that most of the principles still apply.

In short though, you're creating a new thread pool for translations per user request.  This join will block the user's request thread but shouldn't significantly impact other concurrent requests.